src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs

```diff
@@ -0,0 +1,290 @@
+        public bool TryRemove(TKey key, out TValue value)
+        {
+            if (key == null)
+                throw LogHelper.LogArgumentNullException(nameof(key));
+
+            if (!_map.ContainsKey(key))
+            {
+                value = default;
+                return false;
+            }
+
+            // check to make sure node wasn't removed by a different thread
+            if (!_map.TryGetValue(key, out var cacheItem))
+            {
+                value = default;
+                return false;
+            }
+
+            _eventQueue.Add(() => _doubleLinkedList.Remove(cacheItem));
+            _map.TryRemove(key, out _);
+            value = cacheItem.Value;
+
+            return true;
+        }
+
+        /// <summary>
+        /// FOR TESTING ONLY.
+        /// </summary>
+        /// <returns></returns>
+        internal LinkedList<LRUCacheItem<TKey, TValue>> LinkedListValues()
+        {
+            return _doubleLinkedList;
+        }
+
+        /// <summary>
+        /// FOR TESTING ONLY.
+        /// </summary>
+        /// <returns></returns>
+        internal ICollection<LRUCacheItem<TKey, TValue>> MapValues()
+        {
+            return _map.Values;
+        }
+
+        /// <summary>
+        /// Calls <see cref="Dispose(bool)"/> and <see cref="GC.SuppressFinalize"/>
+        /// </summary>
+        public void Dispose()
+        {
+            // Dispose of unmanaged resources.
+            Dispose(true);
+            // Suppress finalization.
+            GC.SuppressFinalize(this);
+        }
+
+        /// <summary>
+        /// If <paramref name="disposing"/> is true, this method disposes of <see cref="_eventQueue"/>.
+        /// </summary>
+        /// <param name="disposing">True if called from the <see cref="Dispose()"/> method, false otherwise.</param>
+        protected virtual void Dispose(bool disposing)
+        {
+            if (!_disposed)
+            {
+                _disposed = true;
+                if (disposing)
+                {
+                    _eventQueue.Dispose();
+                }
+            }
+        }
+    }
+
+
+    /// <summary>
+    /// 
+    /// </summary>
+    /// <typeparam name="TKey"></typeparam>
+    /// <typeparam name="TValue"></typeparam>
+    internal class LRUCacheItem<TKey, TValue>
+    {
+        /// <summary>
+        /// 
+        /// </summary>
+        internal TKey Key { get; set; }
+        /// <summary>
+        /// 
+        /// </summary>
+        internal TValue Value { get; set; }
+        /// <summary>
+        /// 
+        /// </summary>
+        internal DateTime ExpirationTime { get; set; }
+
+        internal LRUCacheItem(TKey key, TValue value)
+        {
+
+            Key = key ?? throw LogHelper.LogArgumentNullException(nameof(key));
+            Value = value ?? throw LogHelper.LogArgumentNullException(nameof(value));
+        }
+
+        internal LRUCacheItem(TKey key, TValue value, DateTime? expirationTime)
```

Comment:
instead of accepting a nullable datetime and then checking if it is null, we can just accept a DateTime and assign it.

Reply:
Similar for Set() above

Reply:
Makes sense! Doesn't really make sense to allow it to be nullable to begin with. Fixed in https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1575/commits/45c168bd46aa16f03a5e5f83c8b9b2b8cef9b070.